src/v/bytes/details/io_fragment.h

```diff
@@ -94,14 +102,265 @@ class io_fragment {
+            return _current == o._current;
+        }
+
+    private:
+        friend class io_fragment_list;
+
+        inline void check_generation() const {
+            dassert(
+              _container->_generation == _my_generation,
+              "Attempting to use an invalidated iterator. The corresponding "
+              "iobuf has been mutated since this iterator was constructed.");
+        }
+
+        pointer _current = nullptr;
+#ifndef NDEBUG
+        size_t _my_generation = 0;
+        container_type* _container = nullptr;
+#endif
+    };
+
+public:
+    io_fragment_list() noexcept = default;
+    io_fragment_list(io_fragment_list&& o) noexcept
+      : _head(std::exchange(o._head, nullptr))
+      , _tail(std::exchange(o._tail, nullptr)) {}
+    io_fragment_list(const io_fragment_list&) = delete;
+    io_fragment_list& operator=(io_fragment_list&&) = delete;
+    io_fragment_list& operator=(const io_fragment_list&) = delete;
+    ~io_fragment_list() noexcept {
+        dassert(
+          _head == nullptr && _tail == nullptr,
+          "io_fragment_list cleanup failure and possible memory leak");
+    };
+
+    using iterator = iter<false, true>;
+    using reverse_iterator = iter<false, false>;
+    using const_iterator = iter<true, true>;
+
+    bool empty() const {
+        check_consistency();
+        return _head == nullptr;
+    }
+
+    io_fragment& front() {
+        check_consistency();
+        return *_head;
+    }
+
+    void pop_front() {
+        check_consistency();
+        io_fragment* h = _head;
+        _head = h->_next;
+        h->_next = nullptr;
+        if (_head != nullptr) {
+            _head->_prev = nullptr;
+        } else {
+            _tail = nullptr;
+        }
+        update_generation();
+    }
+
+    void push_front(io_fragment& elem) {
+        check_consistency();
+        if (_head == nullptr) {
+            _tail = &elem;
+        } else {
+            _head->_prev = &elem;
+        }
+        elem._next = _head;
+        _head = &elem;
+        update_generation();
+    }
+
+    io_fragment& back() {
+        check_consistency();
+        return *_tail;
+    }
+    const io_fragment& back() const {
+        check_consistency();
+        return *_tail;
+    }
+
+    void pop_back() {
+        check_consistency();
+        io_fragment* t = _tail;
+        _tail = _tail->_prev;
+        t->_prev = nullptr;
+        if (_tail != nullptr) {
+            _tail->_next = nullptr;
+        } else {
+            _head = nullptr;
+        }
+        update_generation();
+    }
+
+    void push_back(io_fragment& elem) {
+        assert(elem._prev == nullptr);
+        assert(elem._next == nullptr);
+        check_consistency();
+        if (_tail == nullptr) {
+            _head = &elem;
+        } else {
+            _tail->_next = &elem;
+        }
+        elem._prev = _tail;
+        _tail = &elem;
+        update_generation();
+    }
+
+    template<typename Func>
```

Comment:
Require Func to be noexcept as this method is not exception safe. I.e. if func throws midway _head still points to a disposed memory address.

Reply:
This function is removed in a latter commit